The parameter ``idle_wires`` default flipped to False in all circuit drawers

1ucian0 opened the pull request:
### Summary

Closes #12361 

### Details and comments

As QPUs are getting bigger, also the circuits are growing in terms of qubits.

1ucian0 committed:
The parameter ``idle_wires`` default flipped to False in all circuit drawers

1ucian0 committed:
reno

1ucian0 committed:
readjust tests

1ucian0 committed:
Merge branch 'main' of github.com:Qiskit/qiskit-terra into fixes/12361/1

1ucian0 committed:
test.python.visualization.test_circuit_text_drawer

1ucian0 committed:
adjust visual tests

1ucian0 committed:
Merge branch 'main' of github.com:Qiskit/qiskit-terra into fixes/12361/1

1ucian0 committed:
bug in the tests

1ucian0 committed:
Merge branch 'main' into fixes/12361/1

1ucian0 committed:
adjust test

1ucian0 committed:
black

1ucian0 committed:
Merge branch 'main' of github.com:Qiskit/qiskit-terra into fixes/12361/1

jakelishman commented on releasenotes/notes/closes_12361-d3ea2c442a4a74a7.yaml:
```diff
@@ -0,0 +1,4 @@
+---
+upgrade_visualization:
+  - |
+    The default for the parameter ``idle_wires`` changed to ``False`` in all the circuit drawers. If you still want to see wires without instructions, set ``idle_wires=True`` explicitly. 
```
It might help to either explain a bit more, or put in a couple of text-mode drawings just to illustrate the difference?  I'm not sure users necessarily know what the option is.

1ucian0 replied:
good point. done in 4f83dc702caebb3cb04c909b909d2b7034afd256

jakelishman reviewed (commented):
I was going to say that this is a better default, but then I thought: actually I use the drawers quite a lot for small-scale drawing, and I think suppressing idle wires there by default is quite possibly surprising behaviour.

I wonder if it's worth having `idle_wires` take a number to use as a cut-off size: if there are `n` or more idle wires (possibly only count qubit ones), hide them, and if there's fewer, show them?  So `idle_wires=0` is the same as the current `idle_wires=False`.  We'd have to catch `idle_wires=True` and turn it into `idle_wires=math.inf`, potentially.

Alternatively, we ignore the int stuff, and add a third option to `idle_wires`: `"layout"`.  I think this might be the happiest middle-ground: if the circuit has been transpiled to a layout (so the left margin will indicate qubit mapping), _then_ we set `idle_wires=False`, and if it hasn't, we set `idle_wires=True`.  (Could also call the option `"auto"` or something.)

I'm not 100% convinced of that, I'm just slightly concerned about breaking usability of drawing small-scale circuits - I would think there's a pretty sizeable number of people who almost invariably draw those, and it'd be confusing if wires are sometimes hidden on them.

1ucian0 committed:
testing

1ucian0 committed:
adjust tests

1ucian0 committed:
conflict

1ucian0 committed:
revert test changes

1ucian0 commented on test/visual/mpl/circuit/test_circuit_matplotlib_drawer.py:
```diff
@@ -1117,7 +1117,9 @@ def test_idle_wires_barrier(self):
         fname = "idle_wires_barrier.png"
-        self.circuit_drawer(circuit, output="mpl", cregbundle=False, filename=fname)
+        self.circuit_drawer(
+            circuit, output="mpl", cregbundle=False, filename=fname, idle_wires=False
```
I accidentally found that a test with the description "Test that idle_wires False works with barrier" was not doing `idle_wires=False`

1ucian0 commented on test/visual/mpl/circuit/references/idle_wires_barrier.png:
see https://github.com/Qiskit/qiskit/pull/13865/files#r1977946484

jakelishman commented on qiskit/visualization/circuit/_utils.py:
```diff
@@ -387,7 +387,7 @@ def _get_valid_justify_arg(justify):
 def _get_layered_instructions(
-    circuit, reverse_bits=False, justify=None, idle_wires=True, wire_order=None, wire_map=None
+    circuit, reverse_bits=False, justify=None, idle_wires="auto", wire_order=None, wire_map=None
 ):
```
Can we leave this as `idle_wires: bool`, and resolve the `"auto"` at the top level of the drawers?

jakelishman commented on qiskit/visualization/circuit/circuit_visualization.py:
```diff
@@ -205,7 +207,7 @@ def circuit_drawer(
     default_output = "text"
     default_reverse_bits = False
-    default_idle_wires = config.get("circuit_idle_wires", False)
+    default_idle_wires = config.get("circuit_idle_wires", "auto")
```
Can we now resolve `"auto"` into either `True` or `False` immediately after line 227, so the rest of the visualisers don't need to change?

jakelishman commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -3192,10 +3192,12 @@ def draw(
+                Default is ``"auto"`` unless the
                 user config file (usually ``~/.qiskit/settings.conf``) has an
-                alternative value set. For example, ``circuit_idle_wires = True``.
+                alternative value set. For example, ``circuit_idle_wires = False``.
```
Do we need to update the configuration-file logic to make it possible to specify `"auto"` in there too?

jakelishman reviewed (commented):
Thanks Luciano, this does feel like a good default now.

1ucian0 committed:
new reno

Cryoris committed:
fix layout-related plots

there's some bloch sphere plot failure which likely shouldn't be related to this, let's see if it persists

Cryoris committed:
top-level handling of "auto"

handle "auto" at the circuit_drawer level and subsequently use on booleans in internal logic

jakelishman commented on qiskit/visualization/circuit/circuit_visualization.py:
```diff
@@ -225,6 +225,11 @@ def circuit_drawer(
 
     if idle_wires is None:
         idle_wires = default_idle_wires
+    elif isinstance(idle_wires, str):
```
```suggestion
    if isinstance(idle_wires, str):
```

still need to handle `auto` if `idle_wires` was input as `None` - line 227 would set it to `"auto"` and then it'd skip this block that resolves it

Cryoris replied:
whoops my bad!

Cryoris committed:
Fix idle_wires None case

Co-authored-by: Jake Lishman <jake@binhbar.com>

jakelishman reviewed (approved):
I tried this locally and it seems to do what I expect - thanks for the changes to the default, Luciano, and for getting it over the line, Julien.
